rust/storage/src/network_admission_control.rs

```diff
@@ -0,0 +1,146 @@
+    #[error("IO Error")]
+    IOError,
+    #[error("Error deserializing to block")]
+    DeserializationError,
+}
+
+impl ChromaError for NetworkAdmissionControlError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            NetworkAdmissionControlError::StorageGetError(e) => e.code(),
+            NetworkAdmissionControlError::IOError => ErrorCodes::Internal,
+            NetworkAdmissionControlError::DeserializationError => ErrorCodes::Internal,
+        }
+    }
+}
+
+impl NetworkAdmissionControl {
+    pub fn new(storage: Storage) -> Self {
+        Self {
+            storage,
+            outstanding_requests: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    pub async fn read_from_storage<F, R>(
+        storage: Storage,
+        key: String,
+        f: F,
+    ) -> Result<(), Box<NetworkAdmissionControlError>>
+    where
+        R: Future<Output = Result<(), Box<NetworkAdmissionControlError>>> + Send + 'static,
+        F: (FnOnce(Vec<u8>) -> R) + Send + 'static,
+    {
+        let stream = storage
+            .get(&key)
+            .instrument(tracing::trace_span!(parent: Span::current(), "Storage get"))
+            .await;
+        match stream {
+            Ok(mut bytes) => {
+                let read_block_span =
+                    tracing::trace_span!(parent: Span::current(), "Read bytes to end");
+                let buf = read_block_span
+                    .in_scope(|| async {
+                        let mut buf: Vec<u8> = Vec::new();
+                        while let Some(res) = bytes.next().await {
+                            match res {
+                                Ok(chunk) => {
+                                    buf.extend(chunk);
+                                }
+                                Err(e) => {
+                                    tracing::error!("Error reading from storage: {}", e);
+                                    return Err(Box::new(
+                                        NetworkAdmissionControlError::StorageGetError(e),
+                                    ));
+                                }
+                            }
+                        }
+                        Ok(Some(buf))
+                    })
+                    .await?;
+                let buf = match buf {
+                    Some(buf) => buf,
+                    None => {
+                        // Buffer is empty. Nothing interesting to do.
+                        return Ok(());
+                    }
+                };
+                tracing::info!("Read {:?} bytes from s3", buf.len());
+                return f(buf).await;
+            }
+            Err(e) => {
+                tracing::error!("Error reading from storage: {}", e);
+                return Err(Box::new(NetworkAdmissionControlError::StorageGetError(e)));
+            }
+        }
+    }
+
+    pub async fn get<F, R>(
+        &self,
+        key: String,
+        f: F,
```

Comment:
Is the idea behind allowing user to provide a callback here that there may be logic you don't want to duplicate across duplicate requests? For example block get shouldn't duplicate deserialization and cache add?

Reply:
How do we want to think about hsnw doing range reads ? Should that be application level or transparent in the NAC?